### ISSUE TITLE
"QGDict::hashAsciiKey: Invalid null key" error messages

### ISSUE THREAD

**Issue by jschleus** [doxygen/doxygen#7068]
On a server offering FOSS software and according Doxygen generated source code documentation I found sometimes the message
```
 QGDict::hashAsciiKey: Invalid null key
```
issued into the error channel (often only a handful but for same projects quite a few).

For e.g. for the package `b2evolution` (https://b2evolution.net/media/blogs/downloads/b2evolution-6.11.2-stable-2019-06-13.zip) one got for a nearly native configuration file with the only changes
```
RECURSIVE       = YES
GENERATE_LATEX  = NO
```
982 such messages, for an expanded configuration really used on the server even 25642 ones.

Question: Are that messages harmless or would it be better to exclude such packages from the Doxygen processing since the resulting output may be "falsified"?

I see now that there is already an according pull request #6928.

**Comment by doxygen** [comment]
@jschleus I've just pushed [this commit](https://github.com/doxygen/doxygen/commit/28c10b0c9f4d4ebf1d9ac105241033eaf6fde228) to address the specific issue in b2evolution for the minimal configuration. Please let me know if this also fixes the issue on the real server.

**Comment by albert-github** [comment]
This is a complicated issue I quickly tried 2 packages with the current master and with the version based on #6928 , some initial results:

- b2evolution  
  - current master OK
  - #6928 not OK
- glib-2.61.2 (https://fossies.org/linux/misc/glib-2.61.2.tar.xz)
  - current master NOT OK, 2 messages
  - #6928 OK

There are a number of packages that give similar problems, but this still has to be further investigated.

**Comment by jschleus** [comment]
As addition to Albert's results here results with curent master from the real server for some afftected packages
- b2evolution-6.11.2-stable-2019-06-13 OK (before 25642 messages)
- glib-2.61.2 NOT OK, 2 messages (before 2 messages)
- symfony-4.3.3 OK (before 4 messages)
- opnsense-core-19.7.1 OK (before 8292 messages)

 Comment: The warn log for opnsense-core has now an additional entry  
```
core-19.7.1/src/opnsense/mvc/app/controllers/OPNsense/Diagnostics/Api/SystemhealthController.php:600: warning: Reached end of file while still inside a (nested) comment. Nesting level 1 (probable line reference: 479)
```
Here the link to the file <A HREF="https://github.com/opnsense/core/blob/master/src/opnsense/mvc/app/controllers/OPNsense/Diagnostics/Api/SystemhealthController.php#L477-L598">SystemhealthController.php</A> on GitHub. But I have no idea if that warning is related to the fixed issue!

If I should check more affected packages so let it know me.

**Comment by albert-github** [comment]
I'm currently running all packages where I ever saw the "QGDict::hashAsciiKey: Invalid null key" message. Hopefully I can give an overview tomorrow.

Regarding the message in the warnings log file, I don't think this has to do with the current change.

**Comment by albert-github** [comment]
The results from my tests in respect to the "QGDict::hashAsciiKey: Invalid null key" problem.
I ran the following executables:
- Version 5050a4787b1d55b04bfb6537c0e23ee4de44397b i.e. just before fix 28c10b0c9f4d4ebf1d9ac105241033eaf6fde228
- Version with fix 28c10b0c9f4d4ebf1d9ac105241033eaf6fde228
- Version as of the time of the pull request #6928

The used packages can be found at https://fossies.org/all.html (note this is a floating set of files as newer versions will overwrite older versions).

| Package | Version 5050a4787b1d55b04bfb6537c0e23ee4de44397b i.e. just before fix 28c10b0c9f4d4ebf1d9ac105241033eaf6fde228| Version with fix 28c10b0c9f4d4ebf1d9ac105241033eaf6fde228| pull request #6928|note|
|---------|-------:|----:|-----:|---:|
| OpenSceneGraph-OpenSceneGraph-3.6.4.tar.gz | 2 | 2 | 2 | 4. |
| R-3.6.1.tar.gz | 4 | 4 | 0 | 2. |
| SOGo-4.0.8.tar.gz | 16 | 16 | 0 | 2. |
| b2evolution-6.11.2-stable-2019-06-13.zip | 25654 | 0 | 25792 | 1. |
| cakephp-3.8.1.tar.gz | 52 | 0 | 0 | 3. |
| eventum-3.7.2.tar.xz | 8 | 0 | 0 | 3. |
| evolution-data-server-3.33.4.tar.xz | 12 | 12 | 0 | 2. |
| evolution-ews-3.33.4.tar.xz | 2 | 2 | 0 | 2. |
| gdal-3.0.1.tar.xz | 4 | 4 | 0 | 2. |
| glib-2.61.2.tar.xz | 2 | 2 | 0 | 2. |
| grav-v1.6.11.zip | 4 | 0 | 0 | 3. |
| gretl-2019b.tar.xz | 2 | 2 | 0 | 2. |
| gst-plugins-good-1.16.0.tar.xz | 4 | 4 | 0 | 2. |
| mantisbt-2.21.1.tar.gz | 1 | 1 | 0 | 2. |
| moodle-3.7.1.tgz | 17 | 17 | 0 | 2. |
| moodle-latest-36.tgz | 17 | 17 | 0 | 2. |
| neovim-0.3.7.tar.gz | 1 | 1 | 0 | 2. |
| opnsense-core-19.7.tar.gz | 8292 | 0 | 0 | 3. |
| owncloud-10.2.1.tar.bz2 | 2 | 2 | 0 | 2. |
| pandorafms_console-7.0NG.736.tar.gz | 7 | 7 | 0 | 2. |
| stella-6.0.1.tar.xz | 8 | 8 | 0 | 2. |
| symfony-4.3.3.tar.gz | 4 | 0 | 0 | 2. |
| texstudio-2.12.16.tar.gz | 3 | 3 | 3 | 4. |
| tiki-20.0.tar.bz2 | 4 | 4 | 0 | 2. |
| tine20-allinone_2019.02.8.tar.bz2 | 2 | 2 | 0 | 2. |
| typo3_src-9.5.8.tar.gz | 4 | 4 | 0 | 2. |
| upp-x11-src-13068.tar.gz | 4 | 4 | 0 | 2. |
| vanilla-Vanilla_3.1.tar.gz | 30 | 30 | 0 | 2. |
| wildfly-17.0.1.Final-src.tar.gz | 4 | 4 | 0 | 2. |
| yii-advanced-app-2.0.23.tgz | 13218 | 4 | 13174 | 5. |

We see that:

1. packages benefit from the fix but not from the pull request #6928 
2. packages benefit from the pull request #6928 but not from the fix 
3. Packages that look like to benefit from he fix but and from the pull request #6928 but probably there are some fixes in doxygen that already resolved the problem
4. Packages that benefit neither from the fix nor from the pull request #6928
5. Unclear but probably might benefit from the fix as well as the pull request #6928 and some already resolved problems

**Comment by doxygen** [comment]
I've pushed a [new commit](https://github.com/doxygen/doxygen/commit/61a987f886200cc62a786c3bd558f428062cf529) which includes a number of fixes (including some from #6928). Could you recheck what the result is with this fix in place?

**Comment by albert-github** [comment]
I'll have a go at it, will take some time ...

**Comment by albert-github** [comment]
I reran the packages and in the following packages there are still messages:

| Package | Version 5050a4787b1d55b04bfb6537c0e23ee4de44397b i.e. just before fix 28c10b0c9f4d4ebf1d9ac105241033eaf6fde228| Version with fix 28c10b0c9f4d4ebf1d9ac105241033eaf6fde228| pull request #6928 :walking_man | Version with fix 61a987f886200cc62a786c3bd558f428062cf529|
|---------|-------:|----:|-----:|----:|
| OpenSceneGraph-OpenSceneGraph-3.6.4.tar.gz | 2 | 2 | 2 | 2 |
| R-3.6.1.tar.gz | 4 | 4 | 0 | 4 |
| SOGo-4.0.8.tar.gz | 16 | 16 | 0 | 16 |
| gst-plugins-good-1.16.0.tar.xz | 4 | 4 | 0 | 4 |
| stella-6.0.1.tar.xz | 8 | 8 | 0 | 8 |
| texstudio-2.12.16.tar.gz | 3 | 3 | 3 | 3 |

**Comment by doxygen** [comment]
I've pushed a [new commit](https://github.com/doxygen/doxygen/commit/f16a63d6d5d9fd11a2f1c57bdfad7f1d4b175f0e) which should address the remaining issues.

**Comment by albert-github** [comment]
I just reran the 6 cases that still gave some messages, these messages are now gone.

I'm now running all mentioned cases in the tables, just to be sure there is no regression (it is not expected).

**Comment by albert-github** [comment]
Unfortunately, the regression test, with the f16a63d6d5d9fd11a2f1c57bdfad7f1d4b175f0e executable, gave a new problem (looks like Murphy is always present :disappointed: ):
- glib-2.61.2.tar.xz has now 8 times the "QGDict::hashAsciiKey: Invalid null key" message, attached log  and the used Doxyfile ([glib-2.61.2.tar.gz](https://github.com/doxygen/doxygen/files/3456909/glib-2.61.2.tar.gz))

**Comment by albert-github** [comment]
On Fossies a new problem was shown: (version 2a8d6b0) with the package `glibc-2.30.tar.xz`.
(Note the original package contains some heavily, nested, defined code so best is to exclude here for the tests: math/test-tgmath2.c; or set the `-Denlarge_lex_buffers=<size>` for `cmake` to a sufficient large value).

I've been able been able to reduce the problem to one source file and a default Doxyfile, see attachment: [small_glibc-2.30.tar.gz](https://github.com/doxygen/doxygen/files/3463563/small_glibc-2.30.tar.gz)

**Comment by doxygen** [comment]
@albert-github Looks like this header file is more of a template on how to write machine specific headers, i.e. it is not valid C

```c
/* Replace <machine> with "i386" or "mips" or whatever.  */

struct <machine>_thread_state basic;
```

Anyway [this commit](https://github.com/doxygen/doxygen/commit/844ceaf65cb73fc6b1c316389e5427353edc22dd) should address the issue.

**Comment by albert-github** [comment]
Problem is coming from a larger package I didn't look into the build procedure of the package, the comment also indicates that it is a template.

The problem in `glibc-2.30.tar.xz` looks like to be solved, remains the problem in `glib-2.61.2.tar.xz`, I reran with c2ef54d7056574cf7d25d313680c82563fc3bb65 fix but still 8 messages.

**Comment by doxygen** [comment]
[This commit](https://github.com/doxygen/doxygen/commit/f573274a6f92074a2e31ba35662b2b42c4e9e034) should address the issue found in glib.

**Comment by albert-github** [comment]
I just reran `glib-2.61.2.tar.xz` and this looks OK. I'll now rerun tall mentioned cases in the tables, just to be sure there is no regression (results expected to be ready to be reported tomorrow).

**Comment by jschleus** [comment]
Not really new information, but I just want to confirm that also on fossies.org the doxygen runs for `glibc-2.30.tar.xz` and `glib-2.61.2.tar.xz` look now OK.

A big thank-you for your tenacious and successful work while fixing this issue.

**Comment by albert-github** [comment]
Tests on the mentioned packages were successful.
Proposed pul request #6928 has partly been used for this issue and is now closed.

Problem has been fixed in master on github (please don't close the issue as this will be done at the moment of a release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.8.16.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
